Translate otlp metrics to prom metatada

jmichalek132 opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does

Translate OTLP to Metadata and forward it. 
This makes metrics description and type available when ingest OTLP protocol.

#### Which issue(s) this PR fixes or relates to

Fixes #4910

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/mimir?pullRequest=5693) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/mimir?pullRequest=5693) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/mimir?pullRequest=5693) it.</sub>

gouthamve commented:
https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485

1. Enhancement is correct.
2. Yes, but lets do it in a different PR?
3. Going to defer to the maintainers here. But I _think_ so.
4. You can extend it.

jmichalek132 commented:
* Updated tests, please let me know if this is good enough
* Added TODO for exposing addMetricSuffixes to the configuration used [here](https://github.com/grafana/mimir/pull/5693/files#diff-eb47b1b5619b45185602515d8d5274a06451058d43aaa9c85c9fc1a7222635c3R176)
* Waiting for the decision on  feature flag/configuration option to enable this

jmichalek132 commented:
Hi, @gouthamve  @charleskorn could I please bother you for a review?

charleskorn commented on CHANGELOG.md:
```diff
@@ -66,6 +66,7 @@
 * [ENHANCEMENT] Ingester: optimize label-values with matchers call when number of matched series is small. #5600
 * [ENHANCEMENT] Compactor: Delete bucket-index, markers and debug files if there are no blocks left in the bucket index. This cleanup must be enabled by using `-compactor.no-blocks-file-cleanup-enabled` option. #5648
 * [ENHANCEMENT] Ingester: reduce memory usage of active series tracker. #5665
+* [ENHANCEMENT] Distributor: Translate OTLP to Metadata and forward it. This makes metrics description and type available when ingest OTLP protocol. #5693
```
```suggestion
* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. #5693
```

charleskorn commented on integration/e2emimir/client.go:
```diff
@@ -288,6 +290,29 @@ func (c *Client) LabelValuesCardinality(labelNames []string, selector string, li
 
+// GetPrometheusMetadata fetches the metadata from the Prometheus endpoint /api/v1/metadata.
+func (c *Client) GetPrometheusMetadata(metric string, limit int) (*http.Response, error) {
+	body := make(url.Values)
```
Why not just do `url.Values{}` here?

jmichalek132 replied:
Removed

charleskorn commented on integration/e2emimir/client.go:
```diff
@@ -288,6 +290,29 @@ func (c *Client) LabelValuesCardinality(labelNames []string, selector string, li
+		body.Set("limit", strconv.Itoa(limit))
+	}
+
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("http://%s/prometheus/api/v1/metadata", c.querierAddress), strings.NewReader(body.Encode()))
```
Don't the query parameters (`body`) need to be included in the URL, not the request body?

charleskorn commented on integration/e2emimir/client.go:
```diff
@@ -288,6 +290,29 @@ func (c *Client) LabelValuesCardinality(labelNames []string, selector string, li
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+
+	// Execute HTTP request
```
I don't think this comment is necessary.

jmichalek132 replied:
Removed

charleskorn commented on integration/otlp_ingestion_test.go:
```diff
@@ -71,9 +79,34 @@ func TestOTLPIngestion(t *testing.T) {
+
 	// Push series with histograms to Mimir
 	series, expectedVector, _ = generateHistogramSeries("series", now)
-	res, err = c.PushOTLP(series)
+	res, err = c.PushOTLP(series, metadata)
```
Might be worth adding a check that the metadata is stored correctly for this histogram series too.

jmichalek132 replied:
Added.

charleskorn commented on pkg/util/push/otel.go:
```diff
@@ -150,10 +152,40 @@ func OTLPHandler(
+			scopeMetrics := scopeMetricsSlice.At(j)
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(scopeMetrics.Metrics().At(k).Type()),
+					MetricFamilyName: prometheustranslator.BuildCompliantName(scopeMetrics.Metrics().At(k), "", true), // TODO expose addMetricSuffixes in configuration
+					Help:             scopeMetrics.Metrics().At(k).Description(),
+					Unit:             scopeMetrics.Metrics().At(k).Unit(),
```
I'd suggest extracting `scopeMetrics.Metrics().At(k)` out to a variable to reduce the duplication here.

jmichalek132 replied:
Done.

charleskorn commented on pkg/util/push/otel.go:
```diff
@@ -150,10 +152,40 @@ func OTLPHandler(
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(scopeMetrics.Metrics().At(k).Type()),
+					MetricFamilyName: prometheustranslator.BuildCompliantName(scopeMetrics.Metrics().At(k), "", true), // TODO expose addMetricSuffixes in configuration
```
Is this TODO still relevant?

jmichalek132 replied:
Yes as per @gouthamve request, exposing this boolean into configuration should be done in separate PR
https://github.com/grafana/mimir/pull/5693#issuecomment-1669216300
Answering my question
https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485

charleskorn reviewed (commented):
Thanks for putting this together @jmichalek132. 

As you suggested in [your comment on the issue](https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485), I think it would be good to add a configuration flag to allow enabling / disabling this. For now, let's mark it as experimental and default to disabled until we've had a chance to test this at scale.

I'd also like to get a review from someone more familiar with OTLP before we merge, as I'm not an OTLP expert.

jmichalek132 commented:
Is there a way to run the linter that will also apply the formatting?

charleskorn commented on pkg/util/push/otel.go:
```diff
@@ -45,6 +45,7 @@ func OTLPHandler(
 	maxRecvMsgSize int,
 	sourceIPs *middleware.SourceIPExtractor,
 	allowSkipLabelNameValidation bool,
+	EnableOtelMetadataStorage bool,
```
The convention in Go is for parameter names to be camelCase.

```suggestion
	enableOtelMetadataStorage bool,
```

jmichalek132 replied:
fixed

charleskorn commented on pkg/util/push/push_test.go:
Could you please add a test for the case where the feature flag is disabled?

jmichalek132 replied:
Added here https://github.com/grafana/mimir/pull/5693/files#diff-9cfabdded42ea41657229be85019e9543e391f591bc1aaee09fd36d8b83fc218R130 is this okay?

charleskorn commented on pkg/api/api.go:
```diff
@@ -50,6 +50,9 @@ type ConfigHandler func(actualCfg interface{}, defaultCfg interface{}) http.Hand
 type Config struct {
 	SkipLabelNameValidationHeader bool `yaml:"skip_label_name_validation_header_enabled" category:"advanced"`
 
+	// when true the distributor translate OTLP to metadata and forwards it,
```
I don't think this comment is necessary, we have the description on the flag to help people understand what this is for.

jmichalek132 replied:
removed

charleskorn commented on pkg/api/api.go:
```diff
@@ -67,6 +70,7 @@ type Config struct {
 // RegisterFlags adds the flags required to config this to the given FlagSet.
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SkipLabelNameValidationHeader, "api.skip-label-name-validation-header-enabled", false, "Allows to skip label name validation via X-Mimir-SkipLabelNameValidation header on the http write path. Use with caution as it breaks PromQL. Allowing this for external clients allows any client to send invalid label names. After enabling it, requests with a specific HTTP header set to true wil ...
+	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", false, "Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol.")
```
[nit]

```suggestion
	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", false, "If true, store metadata when ingesting metrics via OTLP. This makes metric descriptions and types available for metrics ingested via OTLP.")
```

charleskorn commented on pkg/util/push/otel.go:
```diff
@@ -150,10 +153,43 @@ func OTLPHandler(
 }
 
+func otelMetricsToMetadata(md pmetric.Metrics) []*mimirpb.MetricMetadata {
+	var metadata []*mimirpb.MetricMetadata
```
One small performance improvement here would be to pre-allocate a slice large enough to hold all of the expected metadata for this request.

jmichalek132 replied:
updated

charleskorn commented:
> Is there a way to run the linter that will also apply the formatting?

`golangci-lint run --fix ./...` will automatically fix some issues.

colega commented:
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

charleskorn commented on pkg/api/api.go:
```diff
@@ -50,6 +50,8 @@ type ConfigHandler func(actualCfg interface{}, defaultCfg interface{}) http.Hand
 type Config struct {
 	SkipLabelNameValidationHeader bool `yaml:"skip_label_name_validation_header_enabled" category:"advanced"`
 
+	enableOtelMetadataStorage bool `yaml:"enable_otel_metadata_translation" category:"experimental"`
```
Could you please add this to the list of experimental features in [the docs](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md)?

jmichalek132 replied:
will do

jmichalek132 replied:
done, is it ok?

charleskorn commented on pkg/util/push/otel.go:
```diff
@@ -150,10 +152,49 @@ func OTLPHandler(
+				metric := scopeMetrics.Metrics().At(k)
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(metric.Type()),
+					MetricFamilyName: prometheustranslator.BuildCompliantName(metric, "", true), // TODO expose addMetricSuffixes in configuration
```
Is this TODO something you're planning to address in this PR?

jmichalek132 replied:
No, @gouthamve asked to leave the implementation of that for follow up PR here https://github.com/grafana/mimir/pull/5693#issuecomment-1669216300, should I remove the comment and create an issue for it?

gouthamve replied:
I'd create an issue yup.

jmichalek132 replied:
Created https://github.com/grafana/mimir/issues/5967. Should I remove the comment or leave it in?

gouthamve replied:
I'd put #5967 in the comment tbh. I'd leave it in.

charleskorn commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@
 * [ENHANCEMENT] Query-frontend: add `cortex_query_frontend_enqueue_duration_seconds` metric that records the time taken to enqueue or reject a query request when not using the query-scheduler. #5879
 * [ENHANCEMENT] Expose `/sync/mutex/wait/total:seconds` Go runtime metric as `go_sync_mutex_wait_total_seconds_total` from all components. #5879
 * [ENHANCEMENT] Go: updated to 1.21.1. #5955
+* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. #5693
```
```suggestion
* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. #5693
```

charleskorn replied:
```suggestion
* [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
```

charleskorn reviewed (approved):
LGTM, thanks @jmichalek132!

charleskorn commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@
 * [ENHANCEMENT] Query-frontend: add `cortex_query_frontend_enqueue_duration_seconds` metric that records the time taken to enqueue or reject a query request when not using the query-scheduler. #5879
 * [ENHANCEMENT] Expose `/sync/mutex/wait/total:seconds` Go runtime metric as `go_sync_mutex_wait_total_seconds_total` from all components. #5879
 * [ENHANCEMENT] Go: updated to 1.21.1. #5955
+* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. #5693
```
```suggestion
* [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
```

charleskorn commented:
Could you please fix the linting and test failures @jmichalek132?

jmichalek132 commented:
> Could you please fix the linting and test failures @jmichalek132?

All should be fixed now. Can you please approve the test run again?

charleskorn commented:
@gouthamve would you mind taking another look at this when you get a chance?

gouthamve commented on pkg/util/push/otel.go:
```diff
@@ -150,10 +152,49 @@ func OTLPHandler(
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				metric := scopeMetrics.Metrics().At(k)
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(metric.Type()),
```
I'm confused here. `metric.Type()` returns this: https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pmetric#MetricType

While https://github.com/grafana/mimir/blob/main/pkg/mimirpb/mimir.pb.go#L53-L64 is a different thing. Specifically OTLP has "Sum" while Mimir has "COUNTER". How are we mapping them together?

jmichalek132 replied:
Yeah I will look into how is this handled in https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusexporter

jmichalek132 replied:
Added the function for translating OTEL metric types to Prometheus metric types.

jmichalek132 commented on pkg/util/push/otel.go:
```diff
@@ -150,10 +152,69 @@ func OTLPHandler(
+	case pmetric.MetricTypeExponentialHistogram:
+		return mimirpb.MetricMetadata_MetricType(3)
+	}
+	return mimirpb.MetricMetadata_MetricType(0)
```
Should I add warning / debug log message here when the default case happens?

gouthamve commented on pkg/util/push/otel.go:
```diff
@@ -162,6 +162,26 @@ func OTLPHandler(
 
+func otelMetricTypeToMimirMetricType(otelMetric pmetric.Metric) mimirpb.MetricMetadata_MetricType {
+	switch otelMetric.Type() {
+	case pmetric.MetricTypeGauge:
+		return mimirpb.MetricMetadata_MetricType(2)
+	case pmetric.MetricTypeSum:
+		metricType := mimirpb.MetricMetadata_MetricType(2)
+		if otelMetric.Sum().IsMonotonic() {
+			metricType = mimirpb.MetricMetadata_MetricType(1)
+		}
+		return metricType
+	case pmetric.MetricTypeHistogram:
+		return mimirpb.MetricMetadata_MetricType(3)
+	case pmetric.MetricTypeSummary:
+		return mimirpb.MetricMetadata_MetricType(5)
+	case pmetric.MetricTypeExponentialHistogram:
+		return mimirpb.MetricMetadata_MetricType(3)
+	}
+	return mimirpb.MetricMetadata_MetricType(0)
```
```suggestion
	case pmetric.MetricTypeGauge:
		return mimirpb.GAUGE
	case pmetric.MetricTypeSum:
		metricType := mimirpb.GAUGE
		if otelMetric.Sum().IsMonotonic() {
			metricType = mimirpb.COUNTER
		}
		return metricType
	case pmetric.MetricTypeHistogram:
		return mimirpb.HISTOGRAM
	case pmetric.MetricTypeSummary:
		return mimirpb.SUMMARY
	case pmetric.MetricTypeExponentialHistogram:
		return mimirpb.HISTOGRAM
	}
	return mimirpb.UNKNOWN
```

gouthamve commented on pkg/util/push/push_test.go:
```diff
@@ -74,7 +74,7 @@ func TestHandlerOTLPPush(t *testing.T) {
 		assert.Equal(t, "foo", series[0].Labels[0].Value)
 
 		metadata := request.Metadata
-		assert.Equal(t, mimirpb.MetricMetadata_MetricType(1), metadata[0].GetType())
```
Similar changes here.

jmichalek132 replied:
fixed

gouthamve reviewed (approved):
LGTM! @charleskorn This is ready to go imo!
